### PR TITLE
Add baggage implementation based on the W3C and OpenTelemetry specification

### LINKED ISSUE #1539: TraceState/Baggage fields currently should only accept string-string pairs
Currently they use the `label.KeyValue`. This is out of compliance with the spec.

### PR DESCRIPTION
**Pull request by MrAlias**
## Related Docs

- [W3C baggage specification](https://w3c.github.io/baggage/)
- [OpenTelemetry baggage specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.3.0/specification/baggage/api.md)
- [RFC 2616 section 2.2](https://datatracker.ietf.org/doc/html/rfc2616#section-2.2) (defines a token, i.e. a baggage key)

### Added

- The `Baggage`, `Member`, and `Property` types are added to the `go.opentelemetry.io/otel/baggage` package along with their related functions.
- The new `ContextWithBaggage`, `ContextWithoutBaggage`, and `FromContext` functions were added to the `go.opentelemetry.io/otel/baggage` package. These functions replace the `Set`, `Value`, `ContextWithValue`, `ContextWithoutValue`, and `ContextWithEmpty` functions from that package and directly work with the new `Baggage` type.

### Changed

- The internal `baggage` package is refactored to be the minimal types needed to work with baggage and context. This package was persisted to maintain the OpenTracing bridge's ability to synchronize state with any baggage updates using hooks.

### Removed

- The `Set`, `Value`, `ContextWithValue`, `ContextWithoutValue`, and `ContextWithEmpty` functions in the `go.opentelemetry.io/otel/baggage` package are removed. Handling of baggage is now done using the added `Baggage` type and related context functions (`ContextWithBaggage`, `ContextWithoutBaggage`, and `FromContext`) in that package.

## Related Issues

Resolves #1539

### CODE REVIEW

**Comment by MadVikingGod** [comment]
This is a bikeshed comment, please feel free to disregard it.

The OTEL spec uses the language name/value pair for Member, and metadata for Property.  I'm ok using the W3C names, but that won't be in line with otel.

File: CHANGELOG.md [hunk 1]
```diff
@@ -33,6 +33,9 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   This method returns the number of list-members the `TraceState` holds. (#1937)
 - Creates package `go.opentelemetry.io/otel/exporters/otlp/otlptrace` that defines a trace exporter that uses a `otlptrace.Client` to send data.
   Creates package `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc` implementing a gRPC `otlptrace.Client` and offers convenience functions, `NewExportPipeline` and `InstallNewPipeline`, to setup and install a `otlptrace.Exporter` in tracing .(#1922)
+- The `Baggage`, `Member`, and `Property` types are added to the `go.opentelemetry.io/otel/baggage` package along with their related functions. (TBD)
+- The new `ContextWithBaggage`, `ContextWithoutBaggage`, and `FromContext` functions were added to the `go.opentelemetry.io/otel/baggage` package.
+  These functions replace the `Set`, `Value`, `ContextWithValue`, `ContextWithoutValue`, and `ContextWithEmpty` functions from that package and directly work with the new `Baggage` type. (TBD)
```
Comment (MrAlias) [inline CHANGELOG.md:38]:
```suggestion
- The `Baggage`, `Member`, and `Property` types are added to the `go.opentelemetry.io/otel/baggage` package along with their related functions. (#1967)
- The new `ContextWithBaggage`, `ContextWithoutBaggage`, and `FromContext` functions were added to the `go.opentelemetry.io/otel/baggage` package.
  These functions replace the `Set`, `Value`, `ContextWithValue`, `ContextWithoutValue`, and `ContextWithEmpty` functions from that package and directly work with the new `Baggage` type. (#1967)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -87,6 +90,8 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   The `"go.opentelemetry.io/otel".Tracer` function or a `TracerProvider` should be used to acquire a library specific `Tracer` instead. (#1900)
 - The `http.url` attribute generated by `HTTPClientAttributesFromHTTPRequest` will no longer include username or password information. (#1919)
 - The `IsEmpty` method of the `TraceState` type in the `go.opentelemetry.io/otel/trace` package is removed in favor of using the added `TraceState.Len` method. (#1931)
+- The `Set`, `Value`, `ContextWithValue`, `ContextWithoutValue`, and `ContextWithEmpty` functions in the `go.opentelemetry.io/otel/baggage` package are removed.
+  Handling of baggage is now done using the added `Baggage` type and related context functions (`ContextWithBaggage`, `ContextWithoutBaggage`, and `FromContext`) in that package. (TBD)
```
Comment (MrAlias) [inline CHANGELOG.md:94]:
```suggestion
  Handling of baggage is now done using the added `Baggage` type and related context functions (`ContextWithBaggage`, `ContextWithoutBaggage`, and `FromContext`) in that package. (#1967)
```

**Comment by MrAlias** [comment]
> This is a bikeshed comment, please feel free to disregard it.
> 
> The OTEL spec uses the language name/value pair for Member, and metadata for Property. I'm ok using the W3C names, but that won't be in line with otel.

I don't have strong feelings here. I'll rename.

**Comment by MrAlias** [comment]
> > This is a bikeshed comment, please feel free to disregard it.
> > The OTEL spec uses the language name/value pair for Member, and metadata for Property. I'm ok using the W3C names, but that won't be in line with otel.
> 
> I don't have strong feelings here. I'll rename.

Digging into this I might have stronger feelings.

The validation errors we return reference W3C terms. The OpenTelemetry specification says nothing about these kinds of errors other than the W3C specification needs to be implemented. It seems prudent to give error messages relevant to the error meaning they will be in terms of the W3C.

Additionally, the OpenTelemetry specification terms are only partially prescriptive in the areas it does specify. The name/value paradigm is not specified for Metadata meaning that we would provide an interface with assumed names. This will be confusing if we are saying the name passed does not match the W3C specification, but the specification calls it a key.

There have been prior discussions in the specification SIG about name choices, and the guidance has been to follow it when it makes sense. I don't think it makes sense here, and there are examples of [other SIGs making similar choices](https://github.com/open-telemetry/opentelemetry-java/blob/0b8131c79e47a8d6ad4f6e7d94d2f902238e3a9c/api/all/src/main/java/io/opentelemetry/api/baggage/Baggage.java#L84-L95) and the inconsistency of introducing the terms in some places and not others will lead to confusion. For these reasons I'm going to stick with the current naming.

**Review by MadVikingGod — APPROVED** [review]
Just some minor notes.

Like I said that is a bikeshed comment.  I personally think that we should try to keep to W3C, but that might put us at odds eventually.

File: baggage/baggage.go [hunk 3]
```diff
@@ -12,56 +12,486 @@
...
+	}
+	return p, nil
+}
+
+// validate ensures p conforms to the W3C Baggage specification, returning an
+// error otherwise.
+func (p Property) validate() error {
+	errFunc := func(err error) error {
+		return fmt.Errorf("invalid property: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply baggage/baggage.go:118]: To make the error distinct from the `Member` validation which calls this.

File: baggage/baggage.go [hunk 4]
```diff
@@ -12,56 +12,486 @@
...
+)
+
+var (
+	errInvalidKey      = errors.New("invalid key")
+	errInvalidValue    = errors.New("invalid value")
+	errInvalidProperty = errors.New("invalid baggage list-member property")
+	errInvalidMember   = errors.New("invalid baggage list-member")
+	errMemberNumber    = errors.New("too many list-members in baggage-string")
+	errMemberBytes     = errors.New("list-member too large")
+	errBaggageBytes    = errors.New("baggage-string too large")
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply baggage/baggage.go:54]: I'm fine exporting if we see a need from users here. I didn't want to pollute the API and would prefer to have user need before doing so. As for matching, I figured only the local unit tests would need to do this, but do you think there is already a need that motivate these being exported?

File: baggage/baggage_test.go [hunk 5]
```diff
@@ -15,72 +15,636 @@
...
+	_, err := New(Member{key: ""})
+	assert.ErrorIs(t, err, errInvalidKey)
+}
+
+func key(n int) string {
+	r := []rune("0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = r[rand.Intn(len(r))]
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply baggage/baggage_test.go:211]: I'm not seeding the rng which is going to make these pseudo random numbers deterministic.

I can explicitly add a static seed value to make it clearer what is happening here.

File: baggage/baggage.go [hunk 6]
```diff
@@ -12,56 +12,486 @@
...
+		return fmt.Errorf("invalid property: %w", err)
+	}
+
+	if !keyRe.MatchString(p.key) {
+		return errFunc(fmt.Errorf("%w: %q", errInvalidKey, p.key))
+	}
+	if p.hasValue && !valueRe.MatchString(p.value) {
+		return errFunc(fmt.Errorf("%w: %q", errInvalidValue, p.value))
+	}
+	if !p.hasValue && p.value != "" {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply baggage/baggage.go:126]: Having an empty value is valid according to the W3C specification. The state of `p.hasValue && p.value == ""` should return `nil` like it does below.

File: baggage/baggage.go [hunk 7]
```diff
@@ -12,56 +12,486 @@
...
+func (p Property) String() string {
+	if p.hasValue {
+		return fmt.Sprintf("%s=%v", p.key, p.value)
+	}
+	return p.key
+}
+
+type properties []Property
+
+func fromInternalProperties(iProps []baggage.Property) properties {
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply baggage/baggage.go:155]: This function returns an unexported type of this package _from_ the internal baggage package's type. The internal in this name refers to the internal package.

File: baggage/baggage.go [hunk 8]
```diff
@@ -12,56 +12,486 @@
...
+			}
+			props = append(props, p)
+		}
+		fallthrough
+	case 1:
+		match := keyValueRe.FindStringSubmatch(parts[0])
+		if len(match) != 3 {
+			return Member{}, fmt.Errorf("%w: %q", errInvalidMember, member)
+		}
+		key, value = match[1], match[2]
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply baggage/baggage.go]: Updated, PTAL.

**Review by Aneurysm9 — APPROVED** [review]

**Pull request merged**